src/Rendering/TOC.hs

```diff
@@ -78,11 +77,16 @@ tableOfContentsBy :: (Foldable f, Functor f)
                   -> Diff f ann ann                       -- ^ The diff to compute the table of contents for.
                   -> [(Change, a)]                        -- ^ A list of entries for relevant changed nodes in the diff.
 tableOfContentsBy selector = fromMaybe [] . cata (\ r -> case r of
-  Patch patch -> (pure . patchEntry <$> bicrosswalk selector selector patch) <> bifoldMap fold fold patch <> Just []
+  Patch edit -> (pure . patchEntry <$> select (bimap selector selector edit)) <> bifoldMap fold fold edit <> Just []
   Merge (In (_, ann2) r) -> case (selector (In ann2 r), fold r) of
     (Just a, Just entries) -> Just ((Changed, a) : entries)
     (_     , entries)      -> entries)
-  where patchEntry = patch (Deleted,) (Inserted,) (const (Replaced,))
+  where patchEntry = edit (Deleted,) (Inserted,) (const (Replaced,))
+
+        select = \case
+          Delete  a   -> Delete <$> a
+          Insert    b -> Insert <$> b
+          Compare a b -> liftA2 Compare a b <|> Delete <$> a <|> Insert <$> b
```

Comment:
This is a tiny bit like a `bisequence`, until you get to the `Compare` case—we want to build a `Compare` over both sides if they exist, falling back to the `Delete`/`Insert` if possible, or otherwise just `Nothing`.

This isn’t amenable to `Data.Edit.fromMaybes` (without further contortions) because we have an `Edit` of `Maybe`s and not a pair.